src/storage/src/dedup_pk_cell_based_row_deserializer.rs

```diff
@@ -0,0 +1,249 @@
+// limitations under the License.
+
+use std::ops::Deref;
+
+use risingwave_common::array::Row;
+use risingwave_common::catalog::OrderedColumnDesc;
+use risingwave_common::error::Result;
+use risingwave_common::types::{Datum, VirtualNode};
+use risingwave_common::util::ordered::OrderedRowDeserializer;
+
+use crate::cell_based_row_deserializer::{CellBasedRowDeserializer, ColumnDescMapping};
+
+/// Similar to [`CellBasedRowDeserializer`], but for dedup pk cell encoding.
+#[derive(Clone)]
+pub struct DedupPkCellBasedRowDeserializer<Desc: Deref<Target = ColumnDescMapping>> {
+    pk_deserializer: OrderedRowDeserializer,
+    inner: CellBasedRowDeserializer<Desc>,
+
+    // Maps pk fields with:
+    // 1. same value and memcomparable encoding,
+    // 2. corresponding row positions.
+    pk_to_row_mapping: Vec<Option<usize>>,
+}
+
+impl<Desc: Deref<Target = ColumnDescMapping>> DedupPkCellBasedRowDeserializer<Desc> {
+    /// Create a [`DedupPkCellBasedRowDeserializer`]
+    /// to decode cell based row with dedup pk encoding.
+    pub fn new(column_mapping: Desc, pk_descs: &[OrderedColumnDesc]) -> Self {
+        let (pk_data_types, pk_order_types) = pk_descs
+            .iter()
+            .map(|ordered_desc| {
+                (
+                    ordered_desc.column_desc.data_type.clone(),
+                    ordered_desc.order,
+                )
+            })
+            .unzip();
+        let pk_deserializer = OrderedRowDeserializer::new(pk_data_types, pk_order_types);
+
+        let pk_to_row_mapping = pk_descs
+            .iter()
+            .map(|d| {
+                let column_desc = &d.column_desc;
+                if column_desc.data_type.mem_cmp_eq_value_enc() {
+                    column_mapping
+                        .get(column_desc.column_id)
+                        .map(|(_, index)| index)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let inner = CellBasedRowDeserializer::new(column_mapping);
+        Self {
+            pk_deserializer,
+            inner,
+            pk_to_row_mapping,
+        }
+    }
+
+    /// Places dedupped pk `Datum`s back into `Row`.
+    fn replace_dedupped_datums_into_row(&self, pk_datums: Vec<Datum>, row: Row) -> Row {
+        let Row(mut row_inner) = row;
+        for (pk_idx, datum) in pk_datums.into_iter().enumerate() {
+            if let Some(row_idx) = self.pk_to_row_mapping[pk_idx] {
+                row_inner[row_idx] = datum;
+            }
+        }
+        Row(row_inner)
+    }
+
+    /// Replaces missing `Datum`s in raw deserialized results.
+    fn replace_dedupped_datums(
+        &self,
+        raw_result: Option<(VirtualNode, Vec<u8>, Row)>,
+    ) -> Result<Option<(VirtualNode, Vec<u8>, Row)>> {
+        if let Some((_vnode, pk, row)) = raw_result {
+            let pk_datums = self.pk_deserializer.deserialize(&pk)?;
+            Ok(Some((
+                _vnode,
```

Comment:
```suggestion
                vnode,
```
If a value is actually used, we can remove the underscore.